### PR TITLE
add scc forbidden error debug around scl image pod creation

### PR DESCRIPTION
**Pull request by gabemontero**
See https://github.com/openshift/cluster-samples-operator/pull/242#issuecomment-599706504 and https://bugzilla.redhat.com/show_bug.cgi?id=1817099 around missing SCC / forbidden errors on pod creates of SCL images 

/assign @adambkaplan 
 @deads2k  FYI

### CODE REVIEW

File: test/extended/image_ecosystem/scl.go [hunk 1]
```diff
@@ -21,6 +22,20 @@ func getPodNameForTest(image string, t tc) string {
...
 
+func debugSCCForbidden(err error, oc *exutil.CLI) (string, error) {
+	debugString := ""
+	if err != nil && kapierrors.IsForbidden(err) {
+		sccList, err := oc.AsAdmin().AdminSecurityClient().SecurityV1().SecurityContextConstraints().List(metav1.ListOptions{})
+		if err != nil {
+			return debugString, err
+		}
+		for _, scc := range sccList.Items {
+			debugString = debugString + "\n" + scc.String()
```
Comment (deads2k) [inline test/extended/image_ecosystem/scl.go]: this doesn't work.  We want a json serialization

File: test/extended/image_ecosystem/scl.go [hunk 2]
```diff
@@ -35,6 +50,12 @@ func defineTest(image string, t tc, oc *exutil.CLI) {
 				Image: t.DockerImageReference,
 			})
 			_, err := oc.KubeClient().CoreV1().Pods(oc.Namespace()).Create(pod)
+			if err != nil {
```
Comment (deads2k) [inline test/extended/image_ecosystem/scl.go]: instead of here (or in addition to), add this to the global error handling for failed tests.

**Comment by gabemontero** [comment]
/assign @deads2k 

are these updates of the sort you wanted ?

**Comment by gabemontero** [comment]
bump @deads2k  - is this the form of debug you want?  

fwiw tests are now green

thanks

**Comment by gabemontero** [comment]
a postgresql deploy in image-eco seemed to just run slow and now finish in time

wasn't the scc forbidden error

/retest

**Comment by gabemontero** [comment]
OK got some forbidden scc errors in image eco https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/24703/pull-ci-openshift-origin-master-e2e-gcp-image-ecosystem/174

will dig into it tomorrow

**Comment by gabemontero** [comment]
So, findings on the repro of the intermittent `forbidden: unable to validate against any security context constraint` with SCC dump that I noted in https://github.com/openshift/origin/pull/24703#issuecomment-606305400 :

1) the SCCs from the dump look fine to my uneducated eye.  They were created before the test ran.  The predefined list looks the same as what I see when I list them in my personal cluster, and the yaml contents look approximate 
2) I see some of the namespace_scc_allocation_controller.go errors in kube-controller-manager noted in some of the bz's somewhat close to the time the test ran, but none of them list the namespace the test ran under
3) Based on my exploration of https://github.com/openshift/openshift-apiserver/tree/master/vendor/github.com/openshift/apiserver-library-go/pkg/securitycontextconstraints and https://github.com/openshift/cluster-policy-controller/blob/master/pkg/security/controller/namespace_scc_allocation_controller.go I'm going to add, at least temporarily in a separate commit, Pod and Namespace YAML dumps for the image-eco test case specifically

@adambkaplan @deads2k FYI

**Comment by gabemontero** [comment]
I'm also considering a WaitForNamespaceToHaveSCCAnnotations method before starting the test, if things point down the timing issue path.

**Comment by gabemontero** [comment]
CI build images failed in image-eco

/test e2e-gcp-image-ecosystem

**Comment by gabemontero** [comment]
OK, the failure in https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/24703/pull-ci-openshift-origin-master-e2e-gcp-image-ecosystem/200 with the latest debug has shed some light @adambkaplan @deads2k 

1) the test namesapce does **NOT** have the SCC annotations at the time of failure:

```
Apr  2 20:45:06.516: INFO: SCC debug namespace yaml:
{
    "metadata": {
        "name": "e2e-test-s2i-usage-kxdxp",
        "selfLink": "/api/v1/namespaces/e2e-test-s2i-usage-kxdxp",
        "uid": "9ce022ed-adc1-4ca3-8b27-ab808931028f",
        "resourceVersion": "23711",
        "creationTimestamp": "2020-04-02T20:45:02Z",
        "annotations": {
            "openshift.io/description": "",
            "openshift.io/display-name": "",
            "openshift.io/requester": "e2e-test-s2i-usage-kxdxp-user"
        }
    },
    "spec": {
        "finalizers": [
            "kubernetes"
        ]
    },
    "status": {
        "phase": "Active"
    }
}
```

2) while the test namesapce is question is not listed, the openshift-kube-controller-manager_kube-controller-manager-ci-op-2vnfl-m-1.c.openshift-gce-devel-ci.internal_cluster-policy-controller.log shows that namespace_scc_allocation_controller.go had 22 separate issues around the time of the test case failure updating namespaces, having to compete with other namespace modifiers: `the object has been modified; please apply your changes to the latest version and try again` ... speculatively, I would not be surprised if we are hitting throttling issues, dropped events, backed up queues, etc. with all that contention and the update never happens before the test catches the forbidden error, fails out, and the namespace is cleaned up

So, action plan
1) I'm going to add a `WaitForNamespaceSCCAnnotations` method in the test code, minimally in scl.go, to see if we can wait for the cluster policy operator to finally catch up
2) I'm going to revisit the cluster-policy-controller/kube-controller-manager code and namespace_scc_allocation_controller.go to see about relist intervals, refetching the namesapce right before modifying, as well as adding a confirmation log that a namespace was successfully updated (minimal log files be damned !!!! ) to assist debug in the future

**Comment by gabemontero** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=1820687 has been opened for ^^

**Pull request closed without merging**